src/mscorlib/src/System/Runtime/Loader/AssemblyLoadContext.cs

```diff
@@ -49,38 +73,91 @@ public abstract class AssemblyLoadContext
         internal static extern void InternalStartProfile(string profile, IntPtr ptrNativeAssemblyLoadContext);
 #endif // FEATURE_MULTICOREJIT
 
-        protected AssemblyLoadContext()
+        static AssemblyLoadContext()
         {
-            // Initialize the ALC representing non-TPA LoadContext
-            InitializeLoadContext(false);
+            // We register the cleanup of all AssemblyLoadContext that have not been finalized in the AppContext.Unloading
+            AppContext.Unloading += OnAppContextUnloading;
         }
 
-        internal AssemblyLoadContext(bool fRepresentsTPALoadContext)
+#if FEATURE_COLLECTIBLE_ALC
+        protected AssemblyLoadContext() : this(false, false)
         {
-            // Initialize the ALC representing TPA LoadContext
-            InitializeLoadContext(fRepresentsTPALoadContext);
         }
-        
-        void InitializeLoadContext(bool fRepresentsTPALoadContext)
+
+        protected AssemblyLoadContext(bool isCollectible) : this(false, isCollectible)
```

Comment:
Couldn't you use the same current define `FEATURE_COLLECTIBLE_ALC`? I'm thinking the define would cause the new APIs to throw and disable code specific to collectible ALC except for the public API surface.